app/src/main/java/com/github/libretube/services/PlaylistDownloadEnqueueService.kt

```diff
@@ -0,0 +1,194 @@
+        var alreadyRetriedOnce = false
+
+        while (nextPage != null) {
+            val playlistPage = runCatching {
+                RetrofitInstance.api.getPlaylistNextPage(playlistId, nextPage!!)
+            }.getOrNull()
+
+            if (playlistPage == null && alreadyRetriedOnce) {
+                toastFromMainDispatcher(R.string.unknown_error)
+                onDestroy()
+                return
+            }
+
+            if (playlistPage == null) {
+                // retry if previous attempt failed
+                alreadyRetriedOnce = true
+                continue
+            }
+
+            alreadyRetriedOnce = false
+            enqueueStreams(playlistPage.relatedStreams)
+            nextPage = playlistPage.nextpage
+        }
+    }
+
+    private suspend fun enqueueStreams(streams: List<StreamItem>) {
+        withContext(Dispatchers.Main) {
+            nManager.notify(ENQUEUE_PROGRESS_NOTIFICATION_ID, buildNotification())
+        }
+
+        for (stream in streams) {
+            val videoInfo = runCatching {
+                RetrofitInstance.api.getStreams(stream.url!!.toID())
+            }.getOrNull() ?: continue
+
+            val videoStream = getStream(videoInfo.videoStreams, maxVideoQuality)
+            val audioStream = getStream(videoInfo.audioStreams, maxAudioQuality)
+
+            val downloadData = DownloadData(
+                videoId = stream.url!!.toID(),
+                fileName = videoInfo.title,
+                videoFormat = videoStream?.format,
+                videoQuality = videoStream?.quality,
+                audioFormat = audioStream?.format,
+                audioQuality = audioStream?.quality,
+                audioLanguage = audioLanguage.takeIf {
+                    videoInfo.audioStreams.any { it.audioTrackLocale == audioLanguage }
+                },
+                subtitleCode = captionLanguage.takeIf {
+                    videoInfo.subtitles.any { it.code == captionLanguage }
+                }
+            )
+            DownloadHelper.startDownloadService(this, downloadData)
+
+            amountOfVideosDone++
+            withContext(Dispatchers.Main) {
+                nManager.notify(ENQUEUE_PROGRESS_NOTIFICATION_ID, buildNotification())
+            }
+        }
+
+        if (amountOfVideos == amountOfVideosDone) onDestroy()
+    }
+
+    private fun getStream(streams: List<PipedStream>, maxQuality: Int?): PipedStream? {
+        val maxStreamQuality = maxQuality ?: return null
+
+        // sort streams by their quality/bitrate
+        val sortedStreams = streams
+            .sortedBy { it.quality.getWhileDigit() }
+
+        // return the last item below the maximum quality - or if there's none - the stream with
+        // the lowest quality available
+        return sortedStreams
+            .lastOrNull { it.quality.getWhileDigit()!! <= maxStreamQuality }
+            ?: sortedStreams.firstOrNull()
```

Comment:
Since this is not a number, won't this have issues when sorting?
Eg: 1080p vs 720p.
I'm not sure if Kotlin converts it to a number.

Another thing to note is quality can have FPS, which can also cause issues.

Reply:
The `getWhileDigit` function reads the string as long as the chars are digit, so "1080p" -> 1080 for example. That works fine, even if FPS are provided too.